pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -257,8 +258,14 @@ func (p *Pod) Stop(ctx context.Context, c client.Client, _ []podset.PodSetInfo,
 		podsInGroup = []corev1.Pod{p.pod}
 	}
 
+	// Pods shouldn't be stopped if matching workload is not found,
+	// since stop will delete the pods.
+	if _, finished := p.Finished(); stopReason == jobframework.StopReasonNoMatchingWorkload && !finished {
+		return false, nil
+	}
+
 	for i := range podsInGroup {
-		if podSuspended(&podsInGroup[i]) || !podsInGroup[i].DeletionTimestamp.IsZero() {
+		if !podsInGroup[i].DeletionTimestamp.IsZero() || (stopReason != jobframework.StopReasonWorkloadDeleted && podSuspended(&podsInGroup[i])) {
```

Comment:
Add a comment for this:

If the workload is being deleted, delete even finished Pods.